### PR TITLE
ShapeShifter for YOLO

### PR DESCRIPTION
**Pull request by dxoigmn**
# What does this PR do?

This obseletes #135 and #177.

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] `pytest`
- [ ] `CUDA_VISIBLE_DEVICES=0 python -m mart experiment=CIFAR10_CNN_Adv trainer=gpu trainer.precision=16` reports 70% (21 sec/epoch).
- [ ] `CUDA_VISIBLE_DEVICES=0,1 python -m mart experiment=CIFAR10_CNN_Adv trainer=ddp trainer.precision=16 trainer.devices=2 model.optimizer.lr=0.2 trainer.max_steps=2925 datamodule.ims_per_batch=256 datamodule.world_size=2` reports 70% (14 sec/epoch).

## Before submitting

- [ ] The title is **self-explanatory** and the description **concisely** explains the PR
- [ ] My **PR does only one thing**, instead of bundling different changes together
- [ ] I list all the **breaking changes** introduced by this pull request
- [ ] I have commented my code
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have run pre-commit hooks with `pre-commit run -a` command without errors

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: mart/callbacks/eval_mode.py
Comment (dxoigmn) [inline mart/callbacks/eval_mode.py:1]: Merge this into #132.

File: mart/callbacks/visualizer.py
Comment (dxoigmn) [inline mart/callbacks/visualizer.py:1]: Merge this into #132.

File: mart/nn/nn.py [hunk 1]
```diff
@@ -80,6 +80,12 @@ def parse_sequence(self, sequence):
                 # We can omit the key of _call_with_args_ if it is the only config.
                 module_cfg = {"_call_with_args_": module_cfg}
 
+            # Add support for calling different functions using dot-syntax
```
Comment (dxoigmn) [inline mart/nn/nn.py:83]: Move `_call_` functionality to new PR.

File: mart/datamodules/coco.py [hunk 2]
```diff
@@ -89,3 +95,23 @@ def __getitem__(self, index: int):
...
+
+def yolo_collate_fn(batch):
+    images, targets = collate_fn(batch)
+
+    # Collate images
+    images = default_collate(images)
+
+    # Turn tuple of dicts into dict of tuples
+    new_targets = {k: tuple(t[k] for t in targets) for k in targets[0].keys()}
+    new_targets["list_of_targets"] = targets
```
Comment (dxoigmn) [inline mart/datamodules/coco.py:108]: Would be nice to get rid of this...

Comment (dxoigmn) [reply mart/datamodules/coco.py:108]: Basically, we either need to decide to accept a `List[Tensor]` for `input` and `List[Dict[str, Tensor]]` for `target` or `Tensor` for `input` and `Dict[str, Tensor]` for `target`. The former interface at least allows images with different shapes and sizes even if the underlying network does not.

File: mart/nn/nn.py [hunk 3]
```diff
@@ -308,3 +316,28 @@ def forward(self, *values, weights=None):
...
+
+# FIXME: This must exist already?!
+class TotalVariation(Sum):
+    def forward(self, *values, weights=None):
+        values = [self._total_variation(value) for value in values]
+
+        return super().forward(*values, weights=weights)
+
+    def _total_variation(self, image):
+        return torch.mean(
```
Comment (dxoigmn) [inline mart/nn/nn.py:329]: Mean doesn't do anything and we should probably normalize this by the size of the patch (e.g., C*H*W).
